### PR TITLE
[12.0][FIX] To approve quotation display & Confirm on creation

### PR DESCRIPTION
**Pull request by tschanzt**
- Add the state to_apporve to quotations instead of sales
- dont allow managers to confirm directly when creating
- allow cancelation in to_approve state

### CODE REVIEW

**Review by leemannd — COMMENTED** [review]
Seems legit.
Could you at least add a comment into your PR to tell people with a few details the points you are trying to solve?

**Review by p-tombez — COMMENTED** [review]
Seems legit, but flake8 is not happy

File: sale_double_validation/models/sale.py [hunk 1]
```diff
@@ -17,6 +19,11 @@ def _setup_fields(self):
                 exists = True
         if not exists:
             selection.insert(0, ('to_approve', _('To Approve')))
+        for field in self._fields:
+            act_field = self._fields.get(field)
```
Comment (mpanarin) [inline sale_double_validation/models/sale.py]: why not use `_fields.items()` or `_fields.values()`?

**Review by mpanarin — CHANGES_REQUESTED** [review]

**Comment by tschanzt** [comment]
@mpanarin addressed. also hopefully pep8 is happy now :pray:

**Review by mpanarin — APPROVED** [review]

**Comment by tschanzt** [comment]
@simahawk im looking for reviewers. im sure you're excited to :smile:

**Comment by simahawk** [comment]
@tschanzt sorry, I have no clue :man_shrugging: 
@p-tombez @leemannd ping :)

**Comment by rousseldenis** [comment]
@tschanzt Do you plan to add a few tests to highlight changes ?

**Comment by tschanzt** [comment]
@rousseldenis i didn't till now :smile: , but i can add some.

**Comment by marylla** [comment]
Hi @tschanzt 
I tested your PR in the runbot and everything worked fine so far, though I have some thoughts about it.

I guess, what you want is a real "draft" state which is unfortunately occupied by the quotation state.
But I think the naming is kind of misleading, especially the name of the button "approve order". 
It actually isn't an order yet and it is also not really an approvement...
Since the "save"-button also does the state conversion from "to approve" to "quotation", you actually don't need the "approve order"-button, do you?

File: sale_double_validation/models/sale.py [hunk 2]
```diff
@@ -17,6 +19,12 @@ def _setup_fields(self):
                 exists = True
         if not exists:
             selection.insert(0, ('to_approve', _('To Approve')))
+        for act_field in self._fields.values():
```
Comment (daramousk) [inline sale_double_validation/models/sale.py:22]: Use `self.fields_get(attributes=['readonly'])`

File: sale_double_validation/models/sale.py [hunk 3]
```diff
@@ -38,10 +46,21 @@ def is_to_approve(self):
     @api.model
     def create(self, vals):
         obj = super().create(vals)
-        if obj.is_to_approve():
-            obj.state = 'to_approve'
+        if not obj.is_to_approve():
```
Comment (daramousk) [inline sale_double_validation/models/sale.py:56]: Why was this reversed?

File: sale_double_validation/models/sale.py [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
-from odoo import api, models, _
+from odoo import api, models, fields, _
 from odoo.tools import float_compare
 
 
 class SaleOrder(models.Model):
     _inherit = 'sale.order'
 
+    state = fields.Selection(default='to_approve')
```
Comment (daramousk) [inline sale_double_validation/models/sale.py:10]: Wouldn't this set the default state to `to_approve` for all the quotations/sale orders? Even for those that `is_to_approve() == False`?

**Review by daramousk — COMMENTED** [review]
I am not really sure about this, this does not seem to be a fix, rather a change of functionality. And things are already complicated when we add a new state to a document.

Perhaps it would be best to have the states not changed and have a different way of approving them.

**Comment by tschanzt** [comment]
@marylla it should only be set to draft if double valdiation is not needed otherwise it will stay in 'to_approve'

File: sale_double_validation/models/sale.py (hunk 4 above)
Comment (tschanzt) [reply sale_double_validation/models/sale.py:10]: yes but if is_to_approve() == False it will be moved to draft on the create

**Comment by rousseldenis** [comment]
@tschanzt Do you plan to fix this ?

File: sale_double_validation/models/sale.py (hunk 3 above)
Comment (tschanzt) [reply sale_double_validation/models/sale.py:56]: we want to_approve to be the default state and move to draft if it doesn't need to be approved. Reason is that in draft state there are possibilities to move it to a different state directly from creation form if we have draft as the default

**Comment by tschanzt** [comment]
@rousseldenis im checking on it.

**Comment by tschanzt** [comment]
@rousseldenis i checked on runbot and it was working as i expected it.  But i think the issue might have been that sale manager skip the process, so if you test with admin it will skip it. I added a flag now if sale managers should be allowed to skip the process or if they still should go through it.

File: sale_double_validation/models/sale.py (hunk 4 above)
Comment (rousseldenis) [reply sale_double_validation/models/sale.py:10]: IMHO, changing this default behaviour could lead to errors (e.g.: you can see in tests). This change should be set by option.

**Comment by rousseldenis** [comment]
Closing this as too old. Please feel free to reopen it if needed.

**Pull request closed without merging**